src/app/router.js

```diff
@@ -0,0 +1,160 @@
+define([
+    'app/config',
+
+    'dojo/_base/lang',
+    'dojo/Deferred',
+    'dojo/hash',
+    'dojo/io-query',
+    'dojo/promise/all',
+    'dojo/request',
+    'dojo/string',
+    'dojo/topic',
+
+    'esri/geometry/Extent',
+
+    'lodash/array/difference'
+], function (
+    config,
+
+    lang,
+    Deferred,
+    hash,
+    ioQuery,
+    all,
+    request,
+    dojoString,
+    topic,
+
+    Extent,
+
+    difference
+) {
+    var obj = {
+        // description:
+        //      Parse and update the URL parameters used to set and preserve the state of the application
+
+        // projectIds: String []
+        //      The id or id's of the currently displayed projects
+        projectIds: [],
+
+        startup: function () {
+            // summary:
+            //      spin up and get everything set up
+            console.log('app/router::startup', arguments);
+
+            topic.subscribe('/dojo/hashchange', lang.hitch(this, 'onHashChange'));
+
+            var beginHash = hash();
+            if (beginHash) {
+                this.onHashChange(beginHash);
+            }
+        },
+        onHashChange: function (newHash) {
+            // summary:
+            //      Fires anytime the hash in the URL changes.
+            //      Check for changes to hash props that we care about
+            //      and fire relevant events.
+            // newHash: String
+            console.log('app/router:onHashChange', arguments);
+
+            var newProps = ioQuery.queryToObject(newHash);
+
+            if (!newProps.id) {
+                newProps.id = [];
+            } else {
+                newProps.id = (typeof newProps.id === 'string') ? [newProps.id] : newProps.id;
+            }
+
+            if (difference(newProps.id, this.projectIds).length > 0 ||
+                difference(this.projectIds, newProps.id).length > 0) {
+                this.projectIds = newProps.id;
+                this.onIdsChange(newProps.id);
+            }
+        },
+        setHash: function (props) {
+            // summary:
+            //      updates the URL with the properties
+            // props: Object
+            //      id: String || String[]
+            console.log('app/router:setHash', arguments);
+
+            hash(ioQuery.objectToQuery(props));
+        },
+        onIdsChange: function (newIds) {
+            // summary:
+            //      fires when the project ids in the hash have changed
+            // newIds: String[]
+            console.log('app/router:onIdsChange', arguments);
+
+            topic.publish(config.topics.projectIdsChanged, newIds);
+        },
+        getProjectsWhereClause: function () {
```

Comment:
I'd be OK moving it. Are you suggesting that it's not related to routing? I can see that.